src/radical/pilot/agent/executing/popen.py

```diff
@@ -164,174 +152,271 @@ def _handle_task(self, task):
+        #       This should be changed to more intuitive integers once MongoDB
+        #       is phased out.
+        #
         # prep stdout/err so that we can append w/o checking for None
         task['stdout'] = ''
         task['stderr'] = ''
 
-        cpt = td['cpu_process_type']
-      # gpt = td['gpu_process_type']  # FIXME: use
+        tid  = task['uid']
+        td   = task['description']
+        sbox = task['task_sandbox_path']
 
-        # FIXME: this switch is insufficient for mixed tasks (MPI/OpenMP)
-        if cpt == 'MPI': launcher = self._mpi_launcher
-        else           : launcher = self._task_launcher
+        launcher = self.find_launcher(task)
 
         if not launcher:
-            raise RuntimeError("no launcher (process type = %s)" % cpt)
-
-        self._log.debug("Launching task with %s (%s).",
-                        launcher.name, launcher.launch_command)
+            raise RuntimeError('no launcher found for task %s' % tid)
+
+        self._log.debug('Launching task with %s', launcher.name)
+
+        launch_script = '%s.launch.sh'        % tid
+        exec_script   = '%s.exec.sh'          % tid
+        exec_path     = '$RP_TASK_SANDBOX/%s' % exec_script
+
+        ru.rec_makedir(sbox)
+
+        with open('%s/%s' % (sbox, launch_script), 'w') as fout:
+
+            fout.write(self._header)
+            fout.write(self._separator)
+            fout.write(self._get_rp_env(task))
+            fout.write(self._get_prof('launch_start', tid))
+
+            fout.write(self._separator)
+            fout.write('# change to task sandbox\n')
+            fout.write('cd $RP_TASK_SANDBOX\n')
+
+            fout.write(self._separator)
+            fout.write('# prepare launcher env\n')
+            fout.write(self._get_launch_env(task, launcher))
+
+            fout.write(self._separator)
+            fout.write('# pre-launch commands\n')
+            fout.write(self._get_prof('launch_pre', tid))
+            fout.write(self._get_pre_launch(task, launcher))
+
+            fout.write(self._separator)
+            fout.write('# launch commands\n')
+            fout.write(self._get_prof('launch_submit', tid))
+            fout.write('%s\n' % self._get_launch_cmds(task, launcher, exec_path))
+            fout.write('RP_RET=$?\n')
+            fout.write(self._get_prof('launch_collect', tid))
+
+            fout.write(self._separator)
+            fout.write('# post-launch commands\n')
+            fout.write(self._get_prof('launch_post', tid))
+            fout.write(self._get_post_launch(task, launcher))
+
+            fout.write(self._separator)
+            fout.write(self._get_prof('launch_stop', tid))
+            fout.write('exit $RP_RET\n')
+
+            fout.write(self._separator)
+            fout.write('\n')
+
+        ranks   = task['slots']['ranks']
+        n_ranks = len(ranks)
+
+        with open('%s/%s' % (sbox, exec_script), 'w') as fout:
```

Comment:
sorry for duplication, just want to keep it up - it's important to have only one call for `write` method per file

Reply:
ack, done.